fpdf/html.py

```diff
@@ -280,16 +283,53 @@ def handle_data(self, data):
                 align=self.align[0].upper(),
                 link=self.href,
             )
-        else:
-            data = data.replace("\n", " ")
+        elif self.pre_formatted: #for pre blocks
+            self.pdf.write(self.h, data)
+
+        elif self.follows_fmt_tag: #don't trim leading whitespace if following a format tag
+            data = re.sub("(\s)(\s*)",self.whitespace_repl, data)
             if self.href:
                 self.put_link(data)
             else:
                 if self.heading_level:
                     self.pdf.start_section(data, self.heading_level - 1)
-                LOGGER.debug("write '%s' h=%d", data.replace("\n", "\\n"), self.h)
+                LOGGER.debug("write '%s' h=%d", re.sub("(\s)(\s*)",self.whitespace_repl, data), self.h)
                 self.pdf.write(self.h, data)
+            self.follows_fmt_tag = False
 
+        else:
+            data = re.sub("^\s+",self.leading_whitespace_repl, data)
+            data = re.sub("(\s)(\s*)",self.whitespace_repl, data)
+            if self.href:
+                self.put_link(data)
+            else:
+                if self.heading_level:
+                    self.pdf.start_section(data, self.heading_level - 1)
+                LOGGER.debug("write '%s' h=%d", re.sub("(\s)(\s*)",self.whitespace_repl, data), self.h)
+                self.pdf.write(self.h, data)
+    
+    def leading_whitespace_repl(self, matchobj):
```

Comment:
Those 2 methods could be functions defined below the class in this module, they do not need to be defined in this class.
`pylint` will probably also warn about this.